mobile_verifier/src/speedtests_average.rs

```diff
@@ -0,0 +1,463 @@
+
+pub const SPEEDTEST_LAPSE: i64 = 48;
+const MIN_DOWNLOAD: u64 = mbps(30);
+const MIN_UPLOAD: u64 = mbps(2);
+const MAX_LATENCY: u32 = 100;
+pub const MIN_REQUIRED_SAMPLES: usize = 2;
+
+pub type EpochAverages = HashMap<PublicKeyBinary, SpeedtestAverage>;
+
+#[derive(Debug, Clone)]
+pub struct SpeedtestAverage {
+    pub pubkey: PublicKeyBinary,
+    pub window_size: usize,
+    pub upload_speed_avg_bps: u64,
+    pub download_speed_avg_bps: u64,
+    pub latency_avg_ms: u32,
+    pub validity: SpeedtestAvgValidity,
+    pub reward_multiplier: Decimal,
+}
+
+impl<'a, I: ?Sized> From<&'a I> for SpeedtestAverage
+where
+    &'a I: IntoIterator<Item = &'a Speedtest>,
+{
+    fn from(iter: &'a I) -> Self {
+        let mut id = vec![]; // eww!
+        let mut window_size = 0;
+        let mut sum_upload = 0;
+        let mut sum_download = 0;
+        let mut sum_latency = 0;
+
+        for Speedtest {
+            pubkey, // eww!
+            upload_speed,
+            download_speed,
+            latency,
+            ..
+        } in speedtests_without_lapsed(iter.into_iter(), Duration::hours(SPEEDTEST_LAPSE))
+        {
+            id = pubkey.as_ref().to_vec(); // eww!
+            sum_upload += *upload_speed as u64;
+            sum_download += *download_speed as u64;
+            sum_latency += *latency as u32;
+            window_size += 1;
+        }
+
+        if window_size > 0 {
+            let upload_speed_avg_bps = sum_upload / window_size as u64;
+            let download_speed_avg_bps = sum_download / window_size as u64;
+            let latency_avg_ms = sum_latency / window_size as u32;
+            let validity = validity(
+                window_size as usize,
+                upload_speed_avg_bps,
+                download_speed_avg_bps,
+                latency_avg_ms,
+            );
+            let tier = SpeedtestTier::new(
+                window_size as usize,
+                upload_speed_avg_bps,
+                download_speed_avg_bps,
+                latency_avg_ms,
+            );
+            let reward_multiplier = tier.into_multiplier();
+            SpeedtestAverage {
+                pubkey: id.into(),
+                window_size: window_size as usize,
+                upload_speed_avg_bps,
+                download_speed_avg_bps,
+                latency_avg_ms,
+                validity,
+                reward_multiplier,
+            }
+        } else {
+            SpeedtestAverage {
+                pubkey: id.into(),
+                window_size: 0,
+                upload_speed_avg_bps: sum_upload,
+                download_speed_avg_bps: sum_download,
+                latency_avg_ms: sum_latency,
+                validity: proto::SpeedtestAvgValidity::TooFewSamples,
+                reward_multiplier: Decimal::ZERO,
+            }
+        }
+    }
+}
+
+impl SpeedtestAverage {
+    pub async fn write(
+        &self,
+        filesink: &FileSinkClient,
+        speedtests: Vec<Speedtest>,
```

Comment:
Instead of having speedtests be an argument, it makes more sense to me to include that in `SpeedtestAverage`

Reply:
The averages are utilzed in multiple points in the code, but the only time the source speedtests from which the average was derived is required is at the point of writing out the proto.  Seems overkill to be throwing the vec of speedtests around unnecessarily